File: qabeldropchatexample/src/main/java/de/qabel/qabeldropchatexample/adapter/ContactsAdapter.java
package de.qabel.qabeldropchatexample.adapter;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import de.qabel.IContact;
import de.qabel.qabeldropchatexample.R;


public class ContactsAdapter extends RecyclerView.Adapter<ContactsAdapter.ViewHolder> {
    private List<IContact> mContacts;
    private OnContactSelectedListener onContactSelectedListener;

    public static class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener{
        private TextView mTextView;
        private TextView textViewContactOwner;
        private OnItemClickListener onItemClickListener;
        public ViewHolder(View v, OnItemClickListener onItemClickListener) {
            super(v);
            mTextView = (TextView) v.findViewById(R.id.contact_text);
            mTextView.setOnClickListener(this);
            this.onItemClickListener = onItemClickListener;

            textViewContactOwner = (TextView) v.findViewById(R.id.textViewContactOwner);
            textViewContactOwner.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            onItemClickListener.onItemClick(view, getPosition());
        }
    }

    public ContactsAdapter(final ArrayList<IContact> contacts, OnContactSelectedListener onContactSelectedListener) {
       this.mContacts = contacts;
       this.onContactSelectedListener = onContactSelectedListener;
    }

    @Override
    public ContactsAdapter.ViewHolder onCreateViewHolder(ViewGroup parent,
                                                   int viewType) {
        View v = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.contact_text_view, parent, false);

        ViewHolder vh = new ViewHolder(v, new OnItemClickListener() {
            @Override
            public void onItemClick(View view, int position) {
                onContactSelectedListener.onContactSelected(
                        mContacts.get(position));
            }
        });
        return vh;
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        holder.mTextView.setText(mContacts.get(position).getAlias());
        holder.textViewContactOwner.setText(mContacts.get(position).getContactOwnerAlias());
    }

    @Override
    public int getItemCount() {
        if (mContacts == null) {
            return 0;
        }
        return mContacts.size();
    }

    public interface OnContactSelectedListener {
        void onContactSelected(IContact contact);
    }

    public interface OnItemClickListener {
        void onItemClick(View view, int position);
    }
}

File: qabelbox/src/main/java/de/qabel/qabelbox/contacts/view/adapters/ContactsAdapter.kt
package de.qabel.qabelbox.contacts.view.adapters

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import de.qabel.core.ui.displayName
import de.qabel.qabelbox.R
import de.qabel.qabelbox.contacts.dto.ContactDto
import de.qabel.qabelbox.contacts.view.adapters.ContactsViewHolder
import java.util.*

class ContactsAdapter(val clickListener: (ContactDto) -> Unit,
                      val longClickListener: (ContactDto) -> Boolean) :
        RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    var TYPE_CONTACT = 1;

    interface ListItem {
        val type: Int
    }

    data class ContactListItem(val contact: ContactDto) : ListItem {
        override val type: Int get() = 1;
    }

    data class HeaderListItem(val header: Char) : ListItem {
        override val type: Int get() = 0;
    }

    private var listItems = mutableListOf<ListItem>();
    private var sectionCount = 0;

    fun refresh(data: List<ContactDto>) {
        var sectionName: Char? = null
        sectionCount = 0
        listItems.clear()
        val sorted = data.toSortedSet(Comparator { t, t2 ->
            t.contact.displayName().compareTo(t2.contact.displayName(), true) })

        sorted.forEach { contact ->
            val section = contact.contact.displayName().first().toUpperCase()
            if (sectionName == null || section != sectionName) {
                listItems.add(HeaderListItem(section))
                sectionName = section
                sectionCount.inc()
            }
            listItems.add(ContactListItem(contact))
        }
    }

    override fun getItemViewType(position: Int): Int {
        return listItems[position].type;
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        when (viewType) {
            TYPE_CONTACT -> {
                val v = LayoutInflater.from(parent.context).inflate(R.layout.item_contacts, parent, false)
                return ContactsViewHolder(v, clickListener, longClickListener);
            }
            else -> {
                val v = LayoutInflater.from(parent.context).inflate(R.layout.list_header, parent, false)
                return ContactsViewHolder.ContactHeaderViewHolder(v);
            }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (holder) {
            is ContactsViewHolder -> holder.bindTo((listItems[position] as ContactListItem).contact)
            is ContactsViewHolder.ContactHeaderViewHolder -> holder.bindTo((listItems[position] as HeaderListItem).header)
        }
    }

    fun getContactCount(): Int {
        return itemCount - sectionCount;
    }

    override fun getItemCount(): Int {
        return listItems.size
    }

}
